### PR TITLE
[MetaSchedule] Task Extraction

### PR DESCRIPTION
**Pull request by junrushao1994**
This PR enables meta schedule to extract tasks from the high-level framework, in our particular case, Relay.

A base class `IntegrationContext` is introduced as the context manager for both task extraction (observer mode) and tuned task injection (injector mode):
- Derived class `TaskExtraction` serves as the observer mode to extract tasks from Relay
- Derived class `ApplyHistoryBest` serves as the injector mode to inject the tuned task back to Relay

The primary API for task extraction is `meta_schedule.integration.extract_task`, which returns a list of tasks on both Relay and TIR-level.

The PR only handles task extraction, but doesn't include further work like:
- Multi-dispatching from Relay to TE/TIR
- Deduplication and similarity score calculation of extracted tasks
- Relay TIR operator strategy
- Layout rewriting

CC: @mbrookhart @mbs-octoml @sunggg @zxybazh @Hzfengsy @YuchenJin @altanh

### CODE REVIEW

File: python/tvm/meta_schedule/integration.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        self.__init_handle_by_constructor__(
+            _ffi_api.ExtractedTask,  # type: ignore # pylint: disable=no-member
+            task_name,
+            mod,
+            dispatched,
+        )
+
+
+@register_object("meta_schedule.IntegrationContext")
+class IntegrationContext(Object):
```
[the remark this answers is not in the document]
Comment (junrushao1994) [reply python/tvm/meta_schedule/integration.py:64]: Okay I did my research looking up dictionaries but didn't end up with a better name...so will go with `ContextBase`

[the remark this answers is not in the document]
Comment (junrushao1994) [reply python/tvm/meta_schedule/integration.py:64]: we debated again offload and it seems `MetaScheduleContext` sounds more...conclusive?

File: python/tvm/meta_schedule/integration.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+from tvm.relay import Function as RelayFunc, vm
+from tvm.runtime import NDArray, Object
+from tvm.target import Target
+from tvm.tir import PrimFunc
+
+from . import _ffi_api
+
+
+@register_object("meta_schedule.ExtractedTask")
+class ExtractedTask(Object):
```
[the remark this answers is not in the document]
Comment (junrushao1994) [reply python/tvm/meta_schedule/integration.py:32]: Okay let's revisit it later :-)

File: python/tvm/meta_schedule/integration.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+@register_object("meta_schedule.ApplyHistoryBest")
+class ApplyHistoryBest(IntegrationContext):
+    pass
+
+
+def extract_task(
+    mod: Union[IRModule, RelayFunc],
+    params: Optional[Dict[str, NDArray]],
+    target: Target,
```
[the remark this answers is not in the document]
Comment (junrushao1994) [reply python/tvm/meta_schedule/integration.py:179]: oh actually I think we should deprecate the string representation of Target - will instead remove the union type

File: python/tvm/meta_schedule/integration.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+
+@register_object("meta_schedule.TaskExtraction")
+class TaskExtraction(IntegrationContext):
+    """An integration context for task extraction
+
+    Parameters
+    ----------
+    tasks : List[ExtractedTask]
+        The extracted tasks
```
[the remark this answers is not in the document]
Comment (junrushao1994) [reply python/tvm/meta_schedule/integration.py:162]: but if so, we are actually not documenting this field

[the remark this answers is not in the document]
Comment (junrushao1994) [reply python/tvm/meta_schedule/integration.py:162]: Looks like we can add docstring after each field, according to PEP-257: https://www.python.org/dev/peps/pep-0257/

File: python/tvm/meta_schedule/integration.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    if not isinstance(target, Target):
+        target = Target(target)
+
+    def _func():
+        with env, _autotvm_silencer(), transform.PassContext(
+            config={
+                "relay.backend.use_meta_schedule": True,
+            },
+            disabled_pass={},
+            opt_level=3,
```
[the remark this answers is not in the document]
Comment (junrushao1994) [reply python/tvm/meta_schedule/integration.py:228]: sounds good. just added the following options to the function parameter: opt_level, disable_pass, pass_config

File: src/meta_schedule/integration.cc [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+bool HasOnlyOneFunction(const IRModule& mod) {
+  if (mod->functions.size() != 1) {
+    return false;
+  }
+  for (const auto& kv : mod->functions) {
+    const BaseFunc& func = kv.second;
+    if (!func->IsInstance<typename FunctionType::ContainerType>()) {
+      return false;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (junrushao1994) [reply src/meta_schedule/integration.cc:38]: Yeah it's checked by the caller of this method

File: python/tvm/meta_schedule/integration.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+        Returns
+        -------
+        ctx : Optional[IntegrationContext]
+            The IntegrationContext in the current scope.
+            NullOpt if it's currently not under any IntegrationContext.
+        """
+        return _ffi_api.IntegrationContextCurrent()  # type: ignore # pylint: disable=no-member
+
+    @staticmethod
+    def entry_point(
```
[the remark this answers is not in the document]
Comment (junrushao1994) [reply python/tvm/meta_schedule/integration.py:113]: Sounds good! Let's do the renaming thing

[the remark this answers is not in the document]
Comment (junrushao1994) [reply python/tvm/meta_schedule/integration.py:113]: I changed `query_with_env` to `query_in_with_scope` which sounds more natural

File: include/tvm/meta_schedule/integration.h [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+   * \param mod The high-level IR
+   * \param dispatched A list of low-level IRs that the high-level IR could potentially dispatch to
+   * \return There are different types of the output
+   * 1) NullOpt if there is no feedback hint
+   * 2) tir::PrimFunc if `mod` should be lowered to a PrimFunc
+   * 3) relay::Function if `mod` should be dispatched to BYOC workflow
+   * 4) IRModule for unified dispatch
+   */
+  static Optional<ObjectRef> QueryInWithScope(runtime::String task_name, IRModule mod,
+                                              Optional<Array<IRModule>> dispatched);
```
Comment (comaniac) [inline include/tvm/meta_schedule/integration.h:127]: This name sounds a bit redundant. `QueryInScope` or `QueryWithScope` may be better, but this is just a nit.

**Review by comaniac — APPROVED** [review]
LGTM. Thanks

File: include/tvm/meta_schedule/integration.h (hunk 8 above)
Comment (junrushao1994) [reply include/tvm/meta_schedule/integration.h:127]: Ahhh I wanted to be as accurate as possible, it's querying inside the with-scope. Renaming to `QueryInsideWithScope` and add some code example to explain

**Review by zxybazh — APPROVED** [review]
LGTM. Minor nit.

File: include/tvm/meta_schedule/integration.h [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+class MetaScheduleContextNode : public runtime::Object {
+ public:
+  /*! \brief Default destructor */
+  virtual ~MetaScheduleContextNode() = default;
+  /*!
+   * \brief The entry point of the integration
+   * \param task_name The name of the task
+   * \param mod The high-level IR
+   * \param dispatched A list of low-level IRs that the high-level IR could potentially dispatch to.
+   * NullOpt means the dispatch needs to be done in the context.
```
[the remark this answers is not in the document]
Comment (junrushao1994) [reply include/tvm/meta_schedule/integration.h]: Sounds good! Added the indent as you suggested

**Pull request merged**